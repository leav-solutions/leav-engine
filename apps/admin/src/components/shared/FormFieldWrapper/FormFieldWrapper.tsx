// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import React from 'react';
import {Form, Label} from 'semantic-ui-react';

export interface IFormFieldWrapperProps {
    error?: string;
    children: React.ReactNode;
}

function FormFieldWrapper({error, children}: IFormFieldWrapperProps): JSX.Element {
    const childrenList: React.ReactNode[] = children ? (!Array.isArray(children) ? [children] : children) : [];

    return (
        <Form.Field>
            {childrenList.map((c, i) => React.cloneElement(c as React.ReactElement<any>, {error: !!error, key: i}))}
            {!!error && <Label pointing="above">{typeof error !== 'string' ? JSON.stringify(error) : error}</Label>}
        </Form.Field>
    );
}

export default FormFieldWrapper;
