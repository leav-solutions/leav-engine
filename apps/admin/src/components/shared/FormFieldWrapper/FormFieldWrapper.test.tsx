// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {shallow} from 'enzyme';
import React from 'react';
import {create} from 'react-test-renderer';
import {Form} from 'semantic-ui-react';
import FormFieldWrapper from '../../shared/FormFieldWrapper';

describe('FormFieldWrapper', () => {
    const testInput = <Form.Input />;
    test('Snapshot test', async () => {
        const comp = create(<FormFieldWrapper>{testInput}</FormFieldWrapper>).toJSON();
        expect(comp).toMatchSnapshot();
    });

    test('Display no error if none', async () => {
        const comp = shallow(<FormFieldWrapper>{testInput}</FormFieldWrapper>);
        expect(comp.find('FormInput').prop('error')).toBe(false);
        expect(comp.find('Label').length).toBe(0);
    });

    test('Display error if any', async () => {
        const comp = shallow(<FormFieldWrapper error="ERROR MESSAGE">{testInput}</FormFieldWrapper>);
        expect(comp.find('FormInput').prop('error')).toBe(true);
        expect(comp.find('Label').shallow().text()).toBe('ERROR MESSAGE');
    });
});
