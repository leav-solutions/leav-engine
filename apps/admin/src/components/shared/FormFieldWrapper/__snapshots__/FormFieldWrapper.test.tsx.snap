// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldWrapper Snapshot test 1`] = `
<div
  className="field"
>
  <div
    className="field"
  >
    <div
      className="ui input"
    >
      <input
        aria-describedby={null}
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
</div>
`;
